### PR TITLE
PyTorch distributed checkpoint async save demo

### PR DESCRIPTION
**Pull request by awonak**
Add demo code for running PyTorch distributed async save checkpoints.

Also updates multinode checkpoint demo to specify the checkpoint save path.

Internal issue: [b/369834594](http://b/369834594)

- [x] Tests pass
- [x] Appropriate changes to documentation are included in the PR

### CODE REVIEW

File: demo/lightning/checkpoint/multinode/strategies.py [hunk 1]
```diff
@@ -120,6 +122,50 @@ def load_checkpoint(self, checkpoint_path):
...
+        converted_state.update({
+            f"optimizer_{idx}": optim_state
+            for idx, optim_state in enumerate(
+                checkpoint.pop("optimizer_states", []))
+        })
+
+        # If a previous future exists, wait for checkpointing to finish,
+        # avoiding queuing more then one checkpoint request at a time.
+        if self._checkpoint_future is not None:
+            self._checkpoint_future.result()
```
[the remark this answers is not in the document]
Comment (awonak) [reply demo/lightning/checkpoint/multinode/strategies.py:155]: I previously had a logging statement measuring the duration of the wait for `result()`. It was consistently less than a millisecond, so I didn't think it was too useful, but perhaps it'll still be good to show the checkpoint save was in fact made asynchronously and useful for debugging if this does become a problem somehow. I'll add that back in.

File: dataflux_pytorch/benchmark/checkpointing/multinode/README.md [hunk 2]
```diff
@@ -57,3 +57,20 @@ To run the script use the following command.
 ```shell
 python3 -m dataflux_pytorch.benchmark.checkpointing.multinode.train
 ```
+
+### Distributed Async Checkpoint Save
+
+Additionally, you can make your distributed checkpointing save calls asynchronous by using the `AsyncDatafluxFSDPStrategy`. Under the hood, this will leverage the [torch.distributed.checkpoitn.async_save](https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict_saver.async_save) to launch the checkpoint save in a separate thread.
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/multinode/README.md:63]: Typo in `torch.distributed.checkpoitn.async_save`

File: demo/lightning/checkpoint/multinode/strategies.py [hunk 3]
```diff
@@ -120,6 +123,54 @@ def load_checkpoint(self, checkpoint_path):
         return metadata
 
 
+class AsyncDatafluxFSDPStrategy(DatafluxFSDPStrategy):
```
Comment (MattIrv) [inline demo/lightning/checkpoint/multinode/strategies.py:126]: Recording what we discussed here:

- Consider implementing the teardown method
- See if it's possible to put this logic directly in `DatafluxFSDPStrategy` with just a boolean async flag or something

**Review by MattIrv — APPROVED** [review]

File: demo/lightning/checkpoint/multinode/strategies.py (hunk 3 above)
Comment (awonak) [reply demo/lightning/checkpoint/multinode/strategies.py:126]: Great suggestions. I'm working on those improvements and will update this PR with those changes.

File: dataflux_pytorch/benchmark/checkpointing/multinode/README.md [hunk 4]
```diff
@@ -57,3 +57,20 @@ To run the script use the following command.
...
+
+This separate demo introduces a new env var to increase the number of training steps between checkpoint saves, which results in more training cycles between checkpoint saves, as well as a var to select using distributed `save` or `async_save` behavior.
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/multinode/README.md:65]: We can shorten this to (and also be specific about the var type) "This separate demo introduces a new env var to increase the number of training steps between checkpoint saves, as well as an env var to select distributed `save` or `async_save` behavior.
Comment"

File: dataflux_pytorch/benchmark/checkpointing/multinode/train_async_save.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        )
+
+        init_start_event = torch.cuda.Event(enable_timing=True)
+        init_end_event = torch.cuda.Event(enable_timing=True)
+
+        init_start_event.record()
+        trainer.fit(model, dataloader)
+        init_end_event.record()
+
+        total_secs = init_start_event.elapsed_time(init_end_event) / 1000
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/multinode/train_async_save.py:120]: nit: might name this "total_time"

File: demo/lightning/checkpoint/multinode/train.py [hunk 6]
```diff
@@ -48,6 +50,7 @@ def init_processes():
     os.environ["NODE_RANK"] = str(rank)
 
     configure_master_addr()
+    return rank
```
Comment (jdnurme) [inline demo/lightning/checkpoint/multinode/train.py:53]: So this function sets the environment variable "NODE_RANK". We don't need to return the rank value since we can just grab it from there (and we should probably keep a single source of truth)

Comment (awonak) [reply demo/lightning/checkpoint/multinode/train.py]: I updated the method to remove rank from the function return. I mostly agree with using a single source of truth for the rank value, but I'm trying to keep env var use centralized to help readability / maintainability.

File: dataflux_pytorch/benchmark/checkpointing/multinode/README.md (hunk 4 above)
Comment (awonak) [reply dataflux_pytorch/benchmark/checkpointing/multinode/README.md:65]: Was your suggestion to remove line 63? Or update line 65 to match the updated env vars?

Comment (jdnurme) [reply dataflux_pytorch/benchmark/checkpointing/multinode/README.md]: Match the env vars and remove some unnecessary text.

File: demo/lightning/checkpoint/multinode/train.py (hunk 6 above)
Comment (jdnurme) [reply demo/lightning/checkpoint/multinode/train.py]: Yes, i think for anything that's not enforced by pytorch itself its better to return. We just don't have much control in this instance.

File: demo/lightning/checkpoint/multinode/strategies.py [hunk 7]
```diff
@@ -19,28 +21,49 @@
                                                        GCSDistributedWriter)
 
 
-def save_checkpoint_helper(rank, checkpoint, path, checkpoint_io, writer):
+def checkpoint_helper(checkpoint):
+    """Flatten the model and optimizer states into a new dict.
```
Comment (Yash9060) [inline demo/lightning/checkpoint/multinode/strategies.py:25]: Can you add a comment explaining why we need this ?

File: demo/lightning/checkpoint/multinode/strategies.py [hunk 8]
```diff
@@ -19,28 +21,49 @@
...
         user_agent.add_dataflux_user_agent(self.storage_client)
 
+        # Attributes used for async behavior.
+        self.use_async = use_async
+        self.checkpoint_group = None
+        self._checkpoint_future = None
+
+        if self.use_async:
+            default_ranks = list(range(dist.get_world_size()))
+            self.checkpoint_group = dist.new_group(
```
Comment (Yash9060) [inline demo/lightning/checkpoint/multinode/strategies.py:64]: Can you explain what does checkpoint_group do ? And why is it needed ? I dont see it being consumed here (but it may be used in some other layer)

**Review by Yash9060 — APPROVED** [review]

File: demo/lightning/checkpoint/multinode/strategies.py (hunk 7 above)
Comment (awonak) [reply demo/lightning/checkpoint/multinode/strategies.py]: Fixed the accuracy of this doc string.

File: dataflux_pytorch/benchmark/checkpointing/multinode/README.md [hunk 9]
```diff
@@ -57,3 +57,20 @@ To run the script use the following command.
 ```shell
 python3 -m dataflux_pytorch.benchmark.checkpointing.multinode.train
 ```
+
+### Distributed Async Checkpoint Save
+
+Additionally, you can make your distributed checkpointing save calls asynchronous by initializing `DatafluxFSDPStrategy` with the kwarg `use_async=True`. Under the hood, this will leverage the [torch.distributed.checkpoint.async_save](https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict_saver.async_save) to launch the checkpoint save in a separate thread.
+
+This separate demo introduces a new env var to increase the number of training steps between checkpoint saves, as well as an env var to select distributed save or async_save behavior:
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/multinode/README.md:65]: nit: don't need and underscore with async_save

File: dataflux_pytorch/benchmark/checkpointing/multinode/train_async_save.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    ) -> None:
+        super().__init__(vocab_size)
+        self.optimizer = optimizer
+        self.simulated_workload_duration = simulated_workload_duration
+        self.model = Transformer(vocab_size=vocab_size, nlayers=nlayers)
+
+    def configure_optimizers(self) -> torch.optim.Optimizer:
+        # Use self.trainer.model.parameters so that we can set
+        # use_orig_params=False on the Strategy. Using AdamW also results in a
+        # checkpoint size roughly 20% of used GPU memory.
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/multinode/train_async_save.py:34]: as in 20% of the GPU memory used for the training, or 20% of the GPU memory that is available for use?

**Review by jdnurme — APPROVED** [review]

File: dataflux_pytorch/benchmark/checkpointing/multinode/train_async_save.py (hunk 10 above)
Comment (awonak) [reply dataflux_pytorch/benchmark/checkpointing/multinode/train_async_save.py]: My understanding is that AdamW uses fewer parameters than SGD making it use less GPU memory for training. However, this optimizer choice is irrelevant to the distributed async save demo, so I've removed it.
